### PR TITLE
GA and Beta Promotions

### PR DESCRIPTION
**Pull request by lukesneeringer**
Promoted to GA (1.0.0):

  * Datastore
  * Logging
  * Storage

Promoted to Beta (0.24.0):

  * Language
  * Translate

Everything else gets a bump to 0.24.0, but still alpha.

### CODE REVIEW

File: core/google/cloud/client.py [hunk 1]
```diff
@@ -137,6 +137,14 @@ def __getstate__(self):
             'Clients have non-trivial state that is local and unpickleable.',
         ]))
 
+    def __getstate__(self):
+        """Explicitly state that clients are not pickleable."""
+
+        raise PicklingError('\n'.join([
+            'Pickling client objects is explicitly not supported.',
+            'Clients have non-trivial state that is local and unpickleable.',
+        ]))
```
Comment (dhermes) [inline core/google/cloud/client.py]: This doesn't belong in this PR? Bad rebase?

File: docs/conf.py [hunk 2]
```diff
@@ -69,7 +69,7 @@
 
 # General information about the project.
 project = u'google-cloud'
-copyright = u'2014, Google'
+copyright = u'2014-2017, Google'
```
Comment (dhermes) [inline docs/conf.py]: Just for fun?

File: resource_manager/setup.py [hunk 3]
```diff
@@ -44,18 +44,19 @@
...
         'Programming Language :: Python :: 3.5',
+        'Programming Language :: Python :: 3.6',
         'Topic :: Internet',
     ],
 }
 
 
 REQUIREMENTS = [
-    'google-cloud-core >= 0.23.1, < 0.24dev',
+    'google-cloud-core >= 0.24.1, < 0.25dev',
```
Comment (dhermes) [inline resource_manager/setup.py]: You want `0.24.0`

File: core/google/cloud/client.py (hunk 1 above)
Comment (lukesneeringer) [inline core/google/cloud/client.py]: I am not sure, probably? It is merged regardless, so whatever.

File: docs/conf.py (hunk 2 above)
Comment (lukesneeringer) [inline docs/conf.py]: Just for correctness.

File: core/google/cloud/client.py [hunk 4]
```diff
@@ -132,6 +132,7 @@ def __init__(self, credentials=None, _http=None):
 
     def __getstate__(self):
         """Explicitly state that clients are not pickleable."""
+
```
Comment (dhermes) [inline core/google/cloud/client.py]: This also is an artifact that came from an old branch

File: datastore/setup.py [hunk 5]
```diff
@@ -44,20 +44,21 @@
         'Programming Language :: Python :: 3',
         'Programming Language :: Python :: 3.4',
         'Programming Language :: Python :: 3.5',
+        'Programming Language :: Python :: 3.6',
```
Comment (dhermes) [inline datastore/setup.py]: @lukesneeringer You wanted this classifier everywhere right? Not just here?

File: bigquery/setup.py [hunk 6]
```diff
@@ -50,12 +50,12 @@
...
 
 REQUIREMENTS = [
-    'google-cloud-core >= 0.23.1, < 0.24dev',
+    'google-cloud-core >= 0.24.0, < 0.25dev',
 ]
 
 setup(
     name='google-cloud-bigquery',
-    version='0.23.0',
+    version='0.24.0',
```
Comment (dhermes) [inline bigquery/setup.py]: `'Programming Language :: Python :: 3.6',` missing in this file

File: core/setup.py [hunk 7]
```diff
@@ -60,7 +60,7 @@
 
 setup(
     name='google-cloud-core',
-    version='0.23.1',
+    version='0.24.0',
```
Comment (dhermes) [inline core/setup.py]: `'Programming Language :: Python :: 3.6',` missing in this file

File: monitoring/setup.py [hunk 8]
```diff
@@ -44,18 +44,19 @@
...
         'Programming Language :: Python :: 3.5',
+        'Programming Language :: Python :: 3.6',
         'Topic :: Internet',
     ],
 }
 
 
 REQUIREMENTS = [
-    'google-cloud-core >= 0.23.1, < 0.24dev',
+    'google-cloud-core >= 0.24.1, < 0.25dev',
```
Comment (dhermes) [inline monitoring/setup.py]: `0.24.0`

File: datastore/setup.py (hunk 5 above)
Comment (lukesneeringer) [inline datastore/setup.py]: Did I miss one?

Comment (dhermes) [inline datastore/setup.py]: Yup. BigQuery and Core

**Comment by dhermes** [comment]
@lukesneeringer There are two "manual" pieces to this release process as well:

- Make a [PR against our HappyBase repo][1]
- Update a [gist][2] that has a "mirror" package for `google-cloud-language` (called `google-cloud-natural-language`) I would really like a less manual process for this but haven't put the effort into figuring one out.

[1]: https://github.com/GoogleCloudPlatform/google-cloud-python-happybase/pull/22
[2]: https://gist.github.com/dhermes/a4e097f2f80fd490b256fc3aa2b4a46b

**Comment by lukesneeringer** [comment]
Great, thanks. I assume you are taking this on once CircleCI passes?

**Comment by dhermes** [comment]
@lukesneeringer Yes I am taking both of those on (once the pushes to PyPI occur, which will require us to push some tags first)

The [latest build][1] for this branch failed because lint tried to grab `google-cloud-core` from PyPI.

ISTM we need to add `session.install('../core/')` to our `lint` sessions?

[1]: https://circleci.com/gh/GoogleCloudPlatform/google-cloud-python/1450

**Comment by dhermes** [comment]
@lukesneeringer I had to push https://github.com/GoogleCloudPlatform/google-cloud-python/pull/3245/commits/84b88781e700e1a1fbb1562ffdc89a8f7b5812de to depend on `logging==1.0.0` in `error_reporting`'s `setup.py` (was previously `0.24.0`).

@tseaver @jonparrott What does a correct upper bound look like there? I went with

```
'google-cloud-logging >= 1.0.0, < 1.1dev',
```

**Comment by jonparrott** [comment]
> What does a correct upper bound look like there? I went with

That works for me.

**Comment by dhermes** [comment]
@daspecster I reverted #3241 and wrapped it into this PR. See my comment on that PR, but the failure seems to contradict your comments about returning a PB.

**Comment by lukesneeringer** [comment]
> @tseaver @jonparrott What does a correct upper bound look like there? I went with

>     'google-cloud-logging >= 1.0.0, < 1.1dev',

I assert it should be `google-cloud-logging >= 1.0.0, < 2.0dev`

**Comment by lukesneeringer** [comment]
Approved jointly by @lukesneeringer and @dhermes.

**Comment by dhermes** [comment]
@lukesneeringer Do we want the https://github.com/GoogleCloudPlatform/google-cloud-python/tree/revert-3245-1.0 branch around still or was that an abandoned idea?

**Comment by lukesneeringer** [comment]
That was a button I clicked accidentally in GitHub. :-)

**Comment by dhermes** [comment]
The "off-build" versions have also been [pushed][1] (the 2nd one was manual):

- https://pypi.python.org/pypi/google-cloud-happybase/0.24.0
- https://pypi.python.org/pypi/google-cloud-natural-language/0.24.0

[1]: https://travis-ci.org/GoogleCloudPlatform/google-cloud-python-happybase/builds/217298729

**Comment by asilversempirical** [comment]
It looks like https://pypi.python.org/pypi/google-cloud-error-reporting was not pushed. Is that intentional, or is it just lagging behind?

**Comment by dhermes** [comment]
@asilversempirical (and @DizzeePascall) I am sorting out an issue with `logging` that is "by accident" blocking `error-reporting` since `logging` is a dependency of `error-reporting`. I hope to sort it out soon.

**Comment by dhermes** [comment]
`error-reporting` is finally going out (as `0.24.0`) and will be pushed by this build:
https://circleci.com/gh/GoogleCloudPlatform/google-cloud-python/1575

Thanks for the patience!

**Comment by pdknsk** [comment]
I think ```storage``` got promoted to GA too early. Many features still missing. Like simultaneous uploading of both _metadata_ and _predefined ACLs_. Or complete support for _versioning_.

**Comment by lukesneeringer** [comment]
GA does not entail feature completion; it simply entails a higher support level and commitment to try and avoid breaking changes.
